sys/net/gnrc/transport_layer/tcp/gnrc_tcp.c

```diff
@@ -0,0 +1,630 @@
+    else{
+        /* Copy Target Address and Port into tcb structure */
+        switch (tcb->address_family) {
+ #ifdef MODULE_GNRC_IPV6
+            case AF_INET6:
+                memcpy(tcb->peer_addr, target_addr, sizeof(ipv6_addr_t));
+                break;
+ #endif
+        }
+        /* Copy Port Information, verfication happens in fsm */
+        *(tcb->local_port) = local_port;
+        *(tcb->peer_port) = target_port;
+
+        /* Setup Timeout: If connection could not be established before */
+        /* the timer expired, the connection attempt failed */
+        connection_timeout_msg.type = MSG_TYPE_CONNECTION_TIMEOUT;
+        xtimer_set_msg(&connection_timeout_timer, GNRC_TCP_CONNECTION_TIMEOUT_DURATION,
+                       &connection_timeout_msg, tcb->owner);
+    }
+
+    /* Call FSM with Event: CALL_OPEN */
+    ret = _fsm(tcb, CALL_OPEN, NULL, NULL, 0);
+    if (ret == -ENOMEM) {
+        DEBUG("gnrc_tcp.c : gnrc_tcp_connect() : Out of receive buffers.\n");
+    } else if(ret == -EADDRINUSE) {
+        DEBUG("gnrc_tcp.c : gnrc_tcp_connect() : local_port is already in use.\n");
+    }
+
+    /* Wait until a connection was established or closed */
+    while (ret >= 0 && tcb->state != CLOSED
+           && tcb->state != ESTABLISHED && tcb->state != CLOSE_WAIT) {
+        msg_receive(&msg);
+        switch (msg.type) {
+            case MSG_TYPE_CONNECTION_TIMEOUT:
+                DEBUG("gnrc_tcp.c : _gnrc_tcp_open() : CONNECTION_TIMEOUT\n");
+                _fsm(tcb, TIMEOUT_CONNECTION, NULL, NULL, 0);
+                ret = -ETIMEDOUT;
+                break;
+
+            case MSG_TYPE_NOTIFY_USER:
+                DEBUG("gnrc_tcp.c : _gnrc_tcp_open() : MSG_TYPE_NOTIFY_USER\n");
+                break;
+
+            default:
+                DEBUG("gnrc_tcp.c : _gnrc_tcp_open() : other message type\n");
+        }
+    }
+
+    /* Cleanup */
+    if (ret != -ETIMEDOUT) {
+      xtimer_remove(&connection_timeout_timer);
+    }
+    if (tcb->state == CLOSED && ret == 0) {
+        ret = -ECONNREFUSED;
+    }
+    tcb->owner = KERNEL_PID_UNDEF;
+    mutex_unlock(&(tcb->function_lock));
+    return ret;
+}
+
+/* External GNRC_TCP API */
+int gnrc_tcp_init(void)
+{
+    /* Guard: Check if thread is already running */
+    if (_gnrc_tcp_pid != KERNEL_PID_UNDEF) {
+        return -1;
+    }
+
+    /* Initialize Mutex for linked-list synchronization */
+    mutex_init(&(_list_gnrc_tcp_tcb_lock));
+
+    /* Initialize Linked-List for connection storage */
+    _list_gnrc_tcp_tcb_head = NULL;
+
+    /* Initialize receive buffers */
+    _rcvbuf_init();
+
+    /* Start TCP processing loop */
+    return thread_create(_stack, sizeof(_stack), GNRC_TCP_PRIO, 0, _event_loop, NULL, "gnrc_tcp");
+}
+
+int gnrc_tcp_tcb_init(gnrc_tcp_tcb_t* tcb, const uint8_t address_family,
+                      uint8_t * const local_addr, uint16_t * const local_port,
+                      uint8_t * const peer_addr, uint16_t * const peer_port)
```

Comment:
omit pointer to port, they should be stored directly in `gnrc_tcp_tcb_t`, see my other comment. 

Further it might also be better, to store `local_addr` and `peer_addr` in there, too; init the arrays with max size, that is size for an IPv6 address. That simplifies the call and the user cannot tamper with the addresses behind your back.